### ISSUE TITLE
Integrate Gateway with LSP Spec

### ISSUE THREAD

**Issue by justinmoon** [fedimint/fedimint#3131]
[Here](https://github.com/BitcoinAndLightningLayerSpecs/lsp) is a spec that a bunch of large LSPs like Breez and Block have agreed to build their APIs around.

Integrating this with our lightning gateway could have 2 benefits:
1) It would reduce the burden of managing liquidity - inbound channels could be created automatically from the LSP to us. But outgoing channels would still need to be manually opened.
2) If we use LDK, `gatewayd` could just be one binary. This would make deployments marginally simpler.

After talking to @johncantrell97, there are basically 2 directions here:
1) Wait for LND / Core Lightning to integrate LSP spec and integrate with this in the gateway
2) Add an LDK-based node to gatewayd and use his [ldk-lsp-client](https://github.com/lightningdevkit/ldk-lsp-client) to communicate with the LSP. The only 2 things we would need to use this library for are (1) ask what the fees are and (2) get an SCID from the LSP, which the gateway would include its registration with the federation. We couldn't directly use [ldk-node](https://github.com/lightningdevkit/ldk-node) because it doesn't support and likely won't add HTLC interception, but John said we could start with [ldk-sample](https://github.com/lightningdevkit/ldk-sample) and add whatever we need there.

**Comment by moneyball** [comment]
@justinmoon Can you describe your HTLC interception use case? LDK supports it but as you note LDK Node does not. We're planning to be very conservative expanding the LDK Node API but if there is a strong need by many users for something it will certainly be considered.

**Comment by m1sterc001guy** [comment]
Here's how a fedimint user receives ecash via lightning:

1. Fedimint user (e.g fedimint-cli) constructs a lightning invoice which contains a route hint to the Lightning Gateway, where the short channel id is a fake scid, which is just a number that corresponds to a federation. 
2. An `IncomingContract` is created on the federation, which essentially uploads the (encrypted) preimage to the invoice to the federation's guardians.
3. The Lightning Gateway's lightning node then listens for incoming HTLCs (using HTLC interception) where the short channel id matches the short channel id for a federation that the gateway is registered to. If it finds one, it will look for a preimage for sale in the federation via the `IncomingContract`. The Lightning Gateway can then use its own ecash balance to "buy" the preimage from the federation (which transfers ecash to the end Fedimint user), and the gateway can then complete the payment using the decrypted preimage.

So I believe we use HTLC interception so that the gateway doesn't need to actually custody funds on behalf of Fedimint users, and instead does this swap between a preimage and ecash notes.

We've already used LDK Node a little bit in some of our tests, I think if we wanted to implement the LSP spec it would probably be beneficial to actually use LDK. Gateway operators would then just need to run a gateway and not need a separate lightning node (although some users with existing lightning nodes will likely still want to do that).

**Comment by m1sterc001guy** [comment]
Yes correct, no need to forward the HTLC to the federation. Without HTLC interception though I'm not sure how/when the Lightning Gateway would know when to go buy the preimage. The HTLC scid is how the gateway knows which federation has the preimage. Perhaps you had a different mechanism in mind.

**Comment by m1sterc001guy** [comment]
That sounds like it will work for our use case as well
